src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -177,78 +180,91 @@ private void environmentVariablesUniqueAndSortedPrecondition(
         ExecutePreconditionViolationType.ENVIRONMENT_VARIABLES_NOT_SORTED);
   }
 
-  private void validateAction(Action action) {
-    Digest commandDigest = action.getCommandDigest();
-    ImmutableList.Builder<Digest> inputDigests = new ImmutableList.Builder<>();
-    inputDigests.add(commandDigest);
-    Iterator<Directory> directoriesIterator =
-      createTreeIterator(action.getInputRootDigest(), "");
+  private void validateActionInputDirectory(
+      Directory directory,
+      Stack<Digest> path,
+      Set<Digest> visited,
+      ImmutableSet.Builder<Digest> inputDigests) {
```

Comment:
And this inputDigests set becomes only what we verify against findMissing with, meaning that multiple files may refer to the same content as you've described.